src/recon_buildblock/PLSPrior.cxx

```diff
@@ -0,0 +1,607 @@
+      const int min_y = current_image_estimate[z].get_min_index();
+      const int max_y = current_image_estimate[z].get_max_index();
+
+      for (int y=min_y;y<= max_y;y++)
+        {
+
+
+          const int min_x = current_image_estimate[z][y].get_min_index();
+          const int max_x = current_image_estimate[z][y].get_max_index();
+
+          for (int x=min_x;x<= max_x;x++)
+            {
+
+              if(x+1>max_x || y+1>max_y ||(z+1>max_z && only_2D==0))
+                  continue;
+
+                /* formula:
+                  sum_x,y,z
+                   div * (pet_im_grad[z][y][x]-inner_product[z][y][x]*anatomical_im_grad[z][y][x]/(*get_norm_sptr ())[z][y][x])*
+                   (*kappa_ptr)[z][y][x] /penalty[z][y][x];
+                */
+
+              if(only_2D==1){
+                  (*gradientx_sptr)[z][y][x+1] =
+                     (((*pet_im_grad_x_sptr)[z][y][x+1]-(*anatomical_grad_x_sptr)[z][y][x+1]*(*inner_product_sptr)[z][y][x+1]/
+                          (*get_norm_sptr ())[z][y][x+1])/(*penalty_sptr)[z][y][x+1] -
+                     (((*pet_im_grad_x_sptr)[z][y][x]-(*anatomical_grad_x_sptr)[z][y][x]*(*inner_product_sptr)[z][y][x]/(*get_norm_sptr ())[z][y][x])/
+                      (*penalty_sptr)[z][y][x]) );
+
+                  (*gradienty_sptr)[z][y+1][x] =
+                      (((*pet_im_grad_y_sptr)[z][y+1][x]-(*anatomical_grad_y_sptr)[z][y+1][x]*(*inner_product_sptr)[z][y+1][x]/
+                          (*get_norm_sptr ())[z][y+1][x])/(*penalty_sptr)[z][y+1][x] -
+                      (((*pet_im_grad_y_sptr)[z][y][x]-(*anatomical_grad_y_sptr)[z][y][x]*(*inner_product_sptr)[z][y][x]/(*get_norm_sptr ())[z][y][x])/
+                       (*penalty_sptr)[z][y][x]) );
+              }
+              else{
+
+                  (*gradientx_sptr)[z][y][x+1] =
+                          (((*pet_im_grad_x_sptr)[z][y][x+1]-(*anatomical_grad_x_sptr)[z][y][x+1]*(*inner_product_sptr)[z][y][x+1]/(*get_norm_sptr ())[z][y][x+1])/
+                          (*penalty_sptr)[z][y][x+1] -
+                          ((*pet_im_grad_x_sptr)[z][y][x]-(*anatomical_grad_x_sptr)[z][y][x]*(*inner_product_sptr)[z][y][x]/(*get_norm_sptr ())[z][y][x])/
+                          (*penalty_sptr)[z][y][x]);
+
+                  (*gradienty_sptr)[z][y+1][x] =
+                          (((*pet_im_grad_y_sptr)[z][y+1][x]-(*anatomical_grad_y_sptr)[z][y+1][x]*(*inner_product_sptr)[z][y+1][x]/
+                          (*get_norm_sptr ())[z][y+1][x])/(*penalty_sptr)[z][y+1][x] -
+                          (((*pet_im_grad_y_sptr)[z][y][x]-(*anatomical_grad_y_sptr)[z][y][x]*(*inner_product_sptr)[z][y][x]/(*get_norm_sptr ())[z][y][x])/(*penalty_sptr)[z][y][x]) );
+
+                  (*gradientz_sptr)[z+1][y][x] =
+                      (((*pet_im_grad_z_sptr)[z+1][y][x]-(*anatomical_grad_z_sptr)[z+1][y][x]*(*inner_product_sptr)[z+1][y][x]/
+                          (*get_norm_sptr ())[z][y+1][x])/(*penalty_sptr)[z+1][y][x] -
```

Comment:
Hi,
The index of the normalization matrix in z direction is wrong (line 547). Should be (*get_norm_sptr ())[z+1][y][x] instead?